ignition-server/controllers/tokensecret_controller.go

```diff
@@ -0,0 +1,172 @@
+	ps.RUnlock()
+	return result, ok
+}
+
+func (ps *payloadStore) Delete(key string) {
+	ps.Lock()
+	delete(ps.store, key)
+	ps.Unlock()
+}
+
+func (ps *payloadStore) Store(key string, value []byte) {
+	ps.Lock()
+	ps.store[key] = value
+	ps.Unlock()
+}
+
+// IgnitionProvider can build ignition payload contents
+// for a given release image.
+type IgnitionProvider interface {
+	// GetPayload returns the ignition payload contents for
+	// the provided release. In the future the list of
+	// inputs may expand to incorporate user-provided ignition
+	// overlay contents.
+	GetPayload(ctx context.Context, token *corev1.Secret) ([]byte, error)
+}
+
+// TokenSecretReconciler watches token Secrets
+// and uses an IgnitionProvider to get a payload out them
+// and stores it in the PayloadsStore.
+// A token Secret is by contractual convention:
+// type: Secret
+//   metadata:
+//   annotations:
+// 	   hypershift.io/ignition-config: "true"
+//	 data:
+//     token: <authz token>
+//     release: <release image string>
+type TokenSecretReconciler struct {
+	client.Client
+	IgnitionProvider IgnitionProvider
+}
+
+func tokenSecretAnnotationPredicate(ctx context.Context) predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "updateEvent"), e.ObjectNew, TokenSecretControllerAnnotation)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "createEvent"), e.Object, TokenSecretControllerAnnotation)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "deleteEvent"), e.Object, TokenSecretControllerAnnotation)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "genericEvent"), e.Object, TokenSecretControllerAnnotation)
+		},
+	}
+}
+func (r *TokenSecretReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}).WithEventFilter(tokenSecretAnnotationPredicate(ctx)).
+		Watches(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForObject{}).
+		WithEventFilter(tokenSecretAnnotationPredicate(ctx)).
+		Complete(r)
+}
+
+func processIfMatchesAnnotation(logger logr.Logger, obj client.Object, annotation string) bool {
+	kind := strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind)
+	log := logger.WithValues("namespace", obj.GetNamespace(), kind, obj.GetName())
+
+	if _, ok := obj.GetAnnotations()[annotation]; ok {
+		log.Info("Resource matches label, will attempt to map resource")
+		return true
+	}
+
+	log.Info("Resource does not match label, will not attempt to map resource")
+	return false
+}
+
+func (r *TokenSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
I don't know that we need a finalizer on the secret. Also, aren't the ignition contents considered immutable per secret? If so could the algorithm be as simple as this:

```
secret, err = client.Get(req)
if notFound(err) || secret.IsDeleted {
  cache.Delete(secret)
  return
}
if cache.HasKey(secret.token) {
  // already cached, nothing to do
  return
}
payload = provider.GetPayload(secret.version)
cache.Store(token, payload)
```

The controller model is a bit awkward for this task. You still need to periodically validate all the existing cache entries against the cluster in case a delete event is missed, but for now we can just rely on it and iterate.

Reply:
yes the token: payload pairs are immutable though something needs to account for deleting the tokens from the payload store which does not have a secret anymore, I don't see how to achieve without the finaliser.
`if notFound(err)` then how would you find the token to be deleted from the payloadStore if you don't have the token since the secret does not exist?

Reply:
A couple of ways I can think of (and I'm sure there are others):

1. Make the cache self-expiring with TTLs for entries (using some existing expiration cache like @csrwng mentioned)
2. In the cache itself, in addition to caching the tokens, also cache the names of secrets from which the tokens were derived, and then periodically requeue all the secret names (which will cause them to be cleaned up if they don't exist anymore)

Reply:
At the time you store the payload/token in the cache you have the secret's metadata, so preserving that info seems key to dealing with cache consistency for the long-running process.

Reply:
Updated:
ExpiringCache enables a cache of pairs "token: payload".
Any pair in the cache is expired once entry.expiry time is above the cache ttl.
The expiry time is renewed for an existing value on every Get operation.
Garbage collection of expired values happens on every Get operation.
We set the TTL to 1h above the reconcile resync period so every existing token Secret has the chance to renew their expiry time on the PayloadStore.Get(token) operation while the non exiting ones get eventually garbageCollected.